nixos/modules/services/networking/3proxy.nix

```diff
@@ -0,0 +1,460 @@
+
+      ${concatMapStringsSep "\n" (service: ''
+        auth ${concatStringsSep " " service.auth}
+
+        ${optionalString (cfg.denyPrivate)
+        "deny * * ${optionalList cfg.privateRanges}"}
+
+        ${concatMapStringsSep "\n" (acl:
+          "${acl.rule} ${
+            concatMapStringsSep " " optionalList [
+              acl.users
+              acl.sources
+              acl.targets
+              acl.targetPorts
+            ]
+          }") service.acl}
+
+        maxconn ${toString service.maxConnections}
+
+        ${optionalString (service.extraConfig != null) service.extraConfig}
+
+        ${service.type} -i${toString service.bindAddress} ${
+          optionalString (service.bindPort != null)
+          "-p${toString service.bindPort}"
+        } ${
+          optionalString (service.extraArguments != null) service.extraArguments
+        }
+
+        flush
+      '') cfg.services}
+      ${optionalString (cfg.extraConfig != null) cfg.extraConfig}
+    '');
+in with lib; {
+  options.services._3proxy = {
+    enable = mkEnableOption "3proxy";
+    confFile = mkOption {
+      type = types.nullOr types.path;
+      default = null;
+      example = "/etc/3proxy.conf";
+      description = ''
+        Ignore all other 3proxy options and load configuration from this file.
+      '';
+    };
+    users = mkOption {
+      type = types.listOf (types.submodule {
+        options = {
+          login = mkOption {
+            type = types.str;
+            example = "user";
+            description = ''
+              User login.
+            '';
+          };
+          passwordType = mkOption {
+            type = types.enum [ "CL" "CR" ];
+            default = "CL";
+            example = "CR";
+            description = ''
+              Password representation. The following values are valid:
+
+              <itemizedlist>
+                <listitem><para>
+                  <literal>"CL"</literal>: cleartext password.
+                </para></listitem>
+                <listitem><para>
+                  <literal>"CR"</literal>: crypt password, only MD5 crypt passwords are supported.
+                </para></listitem>
+              </itemizedlist>
+            '';
+          };
+          password = mkOption {
+            type = types.str;
+            example = "$1$lFDGlder$pLRb4cU2D7GAT58YQvY49.";
+            description = ''
+              User password, either cleartext or md5-crypted.
+            '';
+          };
+        };
+      });
+      default = [ ];
+      example = literalExample ''
+        [
+          {
+            login = "admin";
+            passwordType = "CL";
+            password = "bigsecret";
+          }
+          {
+            login = "test";
+            passwordType = "CR";
+            password = "$1$lFDGlder$pLRb4cU2D7GAT58YQvY49.";
+          }
+        ]
+      '';
+      description = ''
+        User list.
+      '';
+    };
+    usersFile = mkOption {
```

Comment:
It would be nice if you could provide either an example of how the file should look, or a link to documentation from upstream on how the file should look.

Reply:
Done.